### DISCUSSION TITLE
Using `__wbindgen_thread_destroy`

### DISCUSSION THREAD

**Discussion by daxpedda** [wasm-bindgen/wasm-bindgen#3063] [Q&A] [answered]
I was attempting to use `__wbindgen_thread_destroy` in workers introduced by https://github.com/rustwasm/wasm-bindgen/pull/2769. I'm using it as the last function before closing the worker.
But I always get `memory access out of bounds`.

I tried it with the parallel raytrace example:
```diff
diff --git a/examples/raytrace-parallel/worker.js b/examples/raytrace-parallel/worker.js
index 26d40c62d..6fa949c3c 100644
--- a/examples/raytrace-parallel/worker.js
+++ b/examples/raytrace-parallel/worker.js
@@ -19,7 +19,8 @@ self.onmessage = event => {
 
   self.onmessage = async event => {
     // This will queue further commands up until the module is fully initialised:
-    await initialised;
+    let wasm = await initialised;
     wasm_bindgen.child_entry_point(event.data);
+    wasm.__wbindgen_thread_destroy();
   };
 };
```

<details>
<summary>Error message</summary>

```
Uncaught (in promise) RuntimeError: memory access out of bounds
    at dlmalloc::dlmalloc::Chunk::size::h8e4c3cbf7bf433fb (wasm.wasm:0xa01cb)
    at dlmalloc::dlmalloc::Dlmalloc<A>::free::hff62ea1a54cee8e2 (wasm.wasm:0x1c619)
    at dlmalloc::Dlmalloc<A>::free::hbc804e690858b2f2 (wasm.wasm:0x87ab9)
    at std::sys::wasm::alloc::<impl core::alloc::global::GlobalAlloc for std::alloc::System>::dealloc::h41c8086127494701 (wasm.wasm:0x797a4)
    at __rdl_dealloc (wasm.wasm:0x86110)
    at __rust_dealloc (wasm.wasm:0xa3e6f)
    at alloc::alloc::dealloc::h9ae13ce8399aec51 (wasm.wasm:0x8e701)
    at __wbindgen_free (wasm.wasm:0x82095)
    at __wbindgen_thread_destroy (wasm.wasm:0x99d9e)
    at self.onmessage (0bcdd051-d87c-4ecd-b9a0-5d861e6511a1:7:10)
$dlmalloc::dlmalloc::Chunk::size::h8e4c3cbf7bf433fb @ wasm.wasm:0xa01cb
$dlmalloc::dlmalloc::Dlmalloc<A>::free::hff62ea1a54cee8e2 @ wasm.wasm:0x1c619
$dlmalloc::Dlmalloc<A>::free::hbc804e690858b2f2 @ wasm.wasm:0x87ab9
$std::sys::wasm::alloc::<impl core::alloc::global::GlobalAlloc for std::alloc::System>::dealloc::h41c8086127494701 @ wasm.wasm:0x797a4
$__rdl_dealloc @ wasm.wasm:0x86110
$__rust_dealloc @ wasm.wasm:0xa3e6f
$alloc::alloc::dealloc::h9ae13ce8399aec51 @ wasm.wasm:0x8e701
$__wbindgen_free @ wasm.wasm:0x82095
$__wbindgen_thread_destroy @ wasm.wasm:0x99d9e
self.onmessage @ 0bcdd051-d87c-4ecd-b9a0-5d861e6511a1:7
```

</details>

I think adding it might make sure that it doesn't break in the future, or if I'm doing something wrong it would show us how to use it properly.

Friendly ping @jrvidal, maybe you can answer this.

**Comment by jrvidal** [comment]
Do you get the errors after the _first_ click/render? Or on subsequent renders?

The latter is expected b/c the code in the raytracer example reuses workers, so there's really not a natural point where a worker is ready to be destroyed.

**Reply by daxpedda** [reply]
Ah, I see I didn't think this through. Let me cobble together an example that actually tests this properly.

**Reply by daxpedda** [reply]
Alright, this should be a simple enough test:
https://github.com/daxpedda/test-wasm-thread

For me this fails with:
`index out of bounds`

**Reply by jrvidal** [reply]
I tried to run it but I wasn't successful. Doesn't it need an `index.html` file? Where is the `spawn.js` file copied over? What is the `http` command you list in your `README`?

**Reply by daxpedda** [reply]
Apologies, I fixed it now.
> Doesn't it need an `index.html` file?

I gitignored it by mistake. It's present now.
> Where is the `spawn.js` file copied over?

Here: 
https://github.com/daxpedda/test-wasm-thread/blob/5f6ba1eb077efa1889ece5cdf4eeb94561cf5390/src/lib.rs#L28
```rust
let script = include_str!("spawn.js");
```
So it's all done for you, no need to manually copy files.
> What is the `http` command you list in your `README`?

I added an explanation in the README file. I'm using this crate here: https://crates.io/crates/https. But feel free to use any other HTTP hosting solution, you just have to host the `web` folder, where the `index.html` file is.

I remembered also that `spawn.js` includes a direct link to the JS shim, so potentially this has to be fixed if you are not hosting on port 8000.

**Reply by daxpedda** [reply]
Thanks @jrvidal, got everything to work actually. I messed up and used my buggy implementation of #3062.
Sorry!

**Comment by daxpedda** [comment, marked as answer]
After debugging my example a bit. I found out that it actually works, I was accidentally using some local build of `wasm-bindgen` where I messed something up related to #3062.
